New mapblocks-for-sending selection algorithm

celeron55 opened the pull request:
This is a rebase of the mapblock transfer implementation I made in 2015 alongside the farmap branch which never got merged: 
https://github.com/minetest/minetest/pull/3502#issuecomment-322311668

This does not contain the farmap implementation.

The point of this is to set up mapblock transfers in such a way that
- Clients manage their memory by requesting a certain amount of map data to be present from the server
- Clients request a certain weighing of MapBlocks and FarBlocks, allowing them to decide how much of each level of detail they want to get from the server (by player preference)
- The server can optimize the send order of a priority queue of mixed level-of-detail map content
- If you stand still, the world will load and then all MapBlock transfers will stop, if there are no changes happening in the world. The client will not delete anything, and the server will not send anything, as they are able to agree on what part of the map the client wants to have in memory.

The client informs the server about its view range and other mapblock transfer related parameters using the added TOSERVER_SET_WANTED_MAP_SEND_QUEUE = 0x54 network packet.

It also allows easily extending to a farblock system allowing radically less level of detail in larger blocks to be sent in far away areas. For any level-of-detail system the server needs to be able to generate queues of blocks of mixed LoD sorted by a weighed priority according to the client's request. The client can, if it wants to, request only plain mapblocks, or only farblocks (i.e. reduced level of detail), or any ratio of those. This has been ripped out of the system that already generated, sent and rendered farblocks succesfully so the concept works.

To-do:
- [x] Clean up: It's pretty well cleaned up now
- [x] Tweaking for cross version network compatibility (looks like it already was set up for it, both ways)
- [x] Testing for stability and performance (IMO it works fine)
- [x] Testing for cross-version network compatibility (could use some testing by others, but in my tests it works both ways)

Requests:
- Please test for performance and stability.

Side note:
The original [block_transfer_v2](https://github.com/celeron55/minetest/compare/block_transfer_v2) branch also contains a system for splitting mapblock drawlist calculation into multiple frames on the client. I left it out from this for now, but that's a very useful piece of code also. In that branch that code uses `struct DrawListUpdate` in clientmap.cpp.

appgurueu commented:
> `TOSERVER_SET_WANTED_MAP_SEND_QUEUE` packet

Might this open up new possibilities for cheating? E. g. having the server send you specific far away mapblocks - currently the best you can do is turn your view range up ridiculously, which is not effective since you can't decide which specific block you want.

celeron55 commented:
This is the current content for TOSERVER_SET_WANTED_MAP_SEND_QUEUE:

```
	s16 radius_map
	s16 radius_far
	f1000 far_weight
	f1000 fov (radians)
	u32 max_total_mapblocks
	u32 max_total_farblocks
	Manual requests:
	u32 len
	for len:
		u8 type // 1=MapBlock, 2=FarBlock
		v3s16 p
```

This actually gives more insight to the server about what's happening than the current method, so the server has more control, not less.

The client can do manual requests in cases where it decides it needs something specific (not sure what the exact situation is), but the server can of course just not send any of it, for example ever at all, or based on distance from the player, or based on something else. Even based on whether the client is suspected of cheating or not. These checks are not in place at the moment and a distance check should be added. The server probably needs a separate maximum send distance for accurate map data (MapBlocks) and one for reduced LoD data (FarBlocks). Tuning these for each server should allow minimizing opportunity for cheating while allowing long view distances - and for those not interested in any of this, these allow also disabling essentially any new features. New features can be disabled either by the server or by the client.

Adding some extra ways to obscure eg. far-away caves might be added in the future. What's needed is to figure out a network protocol which doesn't make it impossible to do this so that it can be implemented. Given the server still has all the authority, I think implementing such logic should go fairly smoothly using the protocol I'm suggesting.

In case of multiple cameras (in the future), I think the server could just split the given allowance between all known cameras active at the client so that the client doesn't have to request anything manually for them. Just divide (area calculated from the radius given by the client) / (number of cameras active), and translate it back to a radius.

celeron55 commented:
Cleaned this up heavily (even split stuff into a new file, server/autosend.h,cpp), fixed some things and added some comments.

Work is still needed on ensuring compatibility between versions.

I thought for a minute about whether this fits on the roadmap. I think this does because it improves support for the data handling of a reduced-level-of-detail rendering method. Looks like at the moment pretty much anything related to rendering matches the roadmap.

Zughy commented:
@celeron55 even if it doesn't fit, you're part of the staff, so it's automatically "Supported by core dev" anyway

celeron55 commented:
I tested this a few hours and it seems to perform as it should - that is, it keeps up with the mesh generator and map generator and keeps doing stuff until it's logical to stop doing stuff.

rubenwardy commented:
I loaded up a local world and set the view range quite high. It loads slower I think:

![image](https://user-images.githubusercontent.com/2122943/184504802-659bb568-2965-4e61-8e33-148cabb15937.png)

rubenwardy commented:
In master, I can get close to the unknown but cannot out fly/fast it. But with this branch, I'm stopped by the unknown

celeron55 commented:
Does everyone agree trying to catch up and collide into ignore nodes is the use case this should to be tuned for?

I think at the moment this is probably tuned better for getting a large view range to load up while moving slowly.

rubenwardy commented:
It should be able to handle both - prioritising nearby blocks before moving to increase loaded area for a large view range. It can also use velocity as a hint. Lots of players fly fast or use carts/vehicles, running into ignore is more annoying than a smaller view range

celeron55 commented:
On a reasonably fast system like mine (as fast as it may have been, it's now 8 years old), I'm not able to reach ignore when flying fast in minetest_game, not even close. I have like 5 to 6 visible mapblocks ahead of me all the time.

If I tweak down the view range down to 40 or 50, then it's easier to get very close to it, but when trying for a minute or two, I touched it only once.

I can slow down the entire thing by using a debug build instead of release. The most noticeable effect this has is it lowers the mapblock mesh generation throughput to about 6 per frame instead of 40 per frame in the release build.

When I do this, I can hit ignore by flying fast. It seems to me the emergequeue gets too large and the emergethread is stuck generating stuff that's completely irrelevant to the player's fast movement.

Just set something like this to solve it. It works for my debug build to solve the fast fly case:
```
emergequeue_limit_generate = 20
emergequeue_limit_diskonly = 40
```

The reason you don't need to lower these on slow systems for the old block transfer algorithm is probably because it sucks so much it can't fill the queue properly. :-)

The emergequeue limits should probably be dynamically set based on some runtime performance metrics instead of being manual settings.

rubenwardy commented:
My laptop has a i7-8550U (4.0GHz). I was running in debug mode, maybe release optimisations make a big difference

celeron55 commented:
Please try changing the emergequeue limits and re-test.

rubenwardy commented:
It runs fine using release mode. I'll try with debug + the emergequeue limits again

celeron55 commented:
I'm currently testing even lower values, and on the fast build even this has good performance:
```
emergequeue_limit_diskonly = 10
emergequeue_limit_generate = 5
```
Why on earth do we have such high limits by default currently?

rubenwardy commented:
Trying a debug build again, I don't notice a huge difference between the old and new emergequeue limits. The map generates much slower (2 blocks away) when flying than master, but I don't run into it this time. And I get faster loading with release mode (morel like 5 blocks away)

rubenwardy commented:
I think our performance goals should be "not running into ignore when flying fast, and having a larger view range during typical gameplay". So if the map loading slower during fast allows for a larger view range during typical gameplay then that's fine

celeron55 commented:
I'll check the special considerations for the player's velocity from the code in current master and see if there's stuff I can backport into this branch. At a quick glance there's some new stuff that didn't get included in the rebase.

celeron55 commented:
About cross-version network compatibility:
- I tested this on a random public server and it worked fine.
- I noticed this actually already has a small piece of compatibility code that allows it to work as a server for older clients.

celeron55 commented:
I'm removing the WIP title now.

So, please review for approval.

This is the prerequisite for me starting some LoD/farblock/farmesh/farmap/whatever-you-like-to-call-it work. I'll start porting the other half of the ancient version of this branch, see how it feels and then go from there. Whether the final solution to LoD is the ancient branch or something else, I think this PR gives a much better base to build on than current master.

lhofhansl commented:
Can we use this in the future to allow the client to do grouping of blocks when generating meshes?
Currently the number of drawcalls is one of the *the* major reasons for relatively low FPS.
(Compiling with MAP_BLOCKSIZE set to 32 easily doubles FPS, due to reduced drawcalls.)

celeron55 commented:
> Can we use this in the future to allow the client to do grouping of blocks when generating meshes? Currently the number of drawcalls is one of the _the_ major reasons for relatively low FPS. (Compiling with MAP_BLOCKSIZE set to 32 easily doubles FPS, due to reduced drawcalls.)

This doesn't affect the client in that regard basically at all. This just builds a base for the server to send lower level-of-detail stuff over the network and thus allows the server to also generate lower level-of-detail stuff to begin with. But this doesn't actually do that. That needs a lot more code.

Optimizing rendering on the client by batching may be a mostly separate issue, but I'll keep it in mind when looking through the rendering of the reduced level of detail stuff. Maybe it can be usefully coupled into the handling of the FarBlocks which I indend to make very large - I think my old code used a size of 8x8x8 MapBlocks for them.

One point of the FarBlock rendering is to also drop support for special material properties like transparency and just render everything with a single material but different textures, for simplicity and speed. This also might help getting large amounts of otherwise normally detailed stuff rendered. I'll look into it when I get to that point.

lhofhansl commented:
For kicks I compiled with MAP_BLOCKSIZE set to 32 and timed world-loading. It is easily 2x (a new world) to 4x (server loading blocks from disk) slower with this change than without. (view range 300)
(I realize that you might not really care about different block sizes, just adding as another data point)

celeron55 commented:
Ok I've now... done some things. I would now describe this as being DAMN FAST, at least on my mid spec computer.

Please test.

celeron55 commented:
Here's me trying to explain this PR on Discord:

Question:

"Is #12690 allowing further view ranges or just reworking the system? Not totally clear from discussion"

Answer:

"It's enabling the server to prioritize mapblocks smartly enough to allow lower level of detail (LoD) data to be added in a future PR that I'll be making based on some of the rest of my old unmerged code.

I'm splitting it in half because otherwise it would be too large for rebasing or reviewing. And it doesn't make any sense for me to begin work with the rest of it if this isn't accepted.

So basically it does nothing but adds more code [and organizes it somewhat better than the original]. Altough you might find some situations where it's marginally faster or smarter. It could also be marginally slower or dumber, depending on the situation. I think the most important change from a player's perspective might be that once it's loaded the map up to the view range or maximum send distance, whichever comes first, it will next start sending stuff outside the FOV. So if you stand still for a good while and then turn around, the map will have already loaded. In current master if you do that the map will only begin to load once you have turned.

With the LoD stuff I will be aiming for view ranges of like 2000 nodes. It's not just about graphics though, the server needs to somehow be able to send something. It really can't generate, load or send all that at node-by-node resolution in the general case, the trickery needs to start from right there. And for that first comes the network. The PR is about the network."

runsy commented:
I did not understood anything. But if my server will go better, go, go, go! I want it :)

lhofhansl commented:
Tested again. With the latest changes I now find it faster (at least in my scenario) than before.

+1

lhofhansl commented:
@Zughy Is that your approval, or you're adding mine? :)

Zughy commented:
@lhofhansl I was adding yours, I don't review PRs :D

lhofhansl commented on src/clientiface.cpp:
```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
-			if (!(isBlockInSight(p, camera_pos, camera_dir, camera_fov,
-						d_blocks_in_sight, &dist) ||
-					(playerspeed.getLength() > 1.0f * BS &&
-					isBlockInSight(p, camera_pos, playerspeeddir, 0.1f,
```
Will we still get the extra blocks in player-move-direction? I put that in a while ago, so that a player does not fall into an ungenerated map. (Especially if the player is looking into a different direction)

celeron55 replied:
I moved that part to an earlier stage in the process. My version changes the entirety of p, camera_pos, camera_dir and camera_fov to be a block ahead and pointing at the movement direction if the player or the object the player is attached to is moving at considerable speed. Only if the player is stationary or roughly walking do they represent the actual position and camera direction of the player. The exact behavior that this should have is a big question mark but it should be fine for the purpose it's serving at the moment.

lhofhansl commented on src/clientiface.cpp:
```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
-				}
+				// This check is disabled because it mis-guesses sea floors to
+				// not be worth transferring to the client, while they are.
+				/*if (d >= 4 && block->getDayNightDiff() == false)
```
We have a config option for that, so that one can tradeoff rendering glitches vs. saving (a pretty decent amount) of block loads.

celeron55 replied:
It's some really old code that was removed or modified in master already, I need to just delete these commented out checks. That part of this PR is actually dead code in how the client and server act - but the way I specify the protocol requires this method of block fetching to be supported on the server. I'm not sure if it's needed at all.

I'm sorry I don't have time to make and test changes to this right now.

lhofhansl replied:
I added that option some time back. So that one can configure how far off is OK to miss bottom-of-the-ocean blocks.
See: e1842ed370f7cab23e460979a08886b5df38e1b3

lhofhansl replied:
We should leave this in, unless the client can do this in a smarter way.

x2048 commented:
Looking at the packet structure, would it make sense to make it as follows?
```
	u8 lod
	s16 radius
	f1000 weight
	f1000 fov (radians)
	u32 max_blocks
	Manual requests:
	u32 len
	for len:
		v3s16 p
```

It does not add much overhead got two-level LOD system but looks more flexible.

sfan5 commented:
Side note: new packets should not use f1000 but f32 for floats, probably just a documentation mistake since the stream operators already default to f32.

sfan5 commented on src/clientiface.cpp:
```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
+
+			// Update our values to these ones, if they are set
+			if (wanted_range > 0 && camera_fov > 0) {
+				radius_map = wanted_range;
```
shouldn't this be capped by `max_block_send_distance` still?

sfan5 commented on src/clientiface.cpp:
```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
+		u32 max_total_mapblocks = 5000; // 5000 is the default client-side setting
+		u32 max_total_farblocks = 0;
+
+		// One version of Minetest supports wanted_range and camera_fov via
+		// PlayerSAO. Try to use them.
```
(all 5.x versions support this)
```suggestion
		// If the SAO exists, get wanted_range and camera_fov from there.
```

sfan5 commented on src/clientiface.cpp:
```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
-			= LIMITED_MAX_SIMULTANEOUS_BLOCK_SENDS;
-	}
+	WMSSuggestion wmss = m_autosend.getNextBlock(emerge);
+	WantedMapSend wms = wmss.wms;
```
```suggestion
	const WantedMapSend &wms = wmss.wms;
```

sfan5 commented on src/clientiface.cpp:
```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
-
-			// Don't send blocks that are currently being transferred
-			if (m_blocks_sending.find(p) != m_blocks_sending.end())
+	for (size_t i=0; i<m_map_send_queue.size(); i++) {
+		const WantedMapSend &wms = m_map_send_queue[i];
```
```suggestion
	for (const auto &wms : m_map_send_queue) {
```

sfan5 commented on src/clientiface.cpp:
```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
+				// block will not be transferred to it no matter how nicely it
+				// asks. Otherwise the requested send queue is going to get
+				// filled up by these and less important blocks further away
+				// that happen to have been already generated will not transfer.
```
so no `continue` here?

also I didn't manage to determine whether these changes reintroduce the following bug (I have a feeling it might): 5430770b6851a18417f6b2629c9167a5aa38baa5

sfan5 commented on src/clientiface.cpp:
```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
+	if (m_blocks_sending.find(wms) != m_blocks_sending.end()) {
+		if (wms.type == WMST_MAPBLOCK)
+			m_mapblocks_sent[wms.p] = m_blocks_sending[wms];
+		else if (wms.type == WMST_MAPBLOCK)
```
copy-paste mistake

sfan5 commented on src/clientiface.cpp:
```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
+	assert(!m_blocks_sending.count(wms) ||
+			m_blocks_updated_while_sending.count(wms) &&
+			"Block shouldn't be re-sent if it is being sent and was not"
+			" updated while being sent");
```
should be a `FATAL_ERROR_IF` or print to `warningstream` depending on conditions

sfan5 commented on src/clientiface.h:
```diff
@@ -250,51 +229,70 @@ class RemoteClient
-	void GotBlock(v3s16 p);
+	// Shall be called every time before starting to ask a bunch of blocks by
+	// calling getNextBlock()
+	void cycleAutosendAlgorithm(float dtime);
```
Unless I misunderstood `step` would be the better terminology here

sfan5 commented on src/clientiface.h:
```diff
@@ -250,51 +229,70 @@ class RemoteClient
 	 */
-	void ResendBlockIfOnWire(v3s16 p);
+	void ResendBlockIfOnWire(const WantedMapSend &wms);
+	void ResendMapBlockIfOnWire(v3s16 p){
+			return ResendBlockIfOnWire(WantedMapSend(WMST_MAPBLOCK, p)); }
```
```suggestion
	inline void ResendMapBlockIfOnWire(v3s16 p) {
		return ResendBlockIfOnWire(WantedMapSend(WMST_MAPBLOCK, p));
	}
```

sfan5 commented on src/map.cpp:
```diff
@@ -305,6 +305,20 @@ bool Map::removeNodeWithEvent(v3s16 p)
+	v3f block_center_p(
+			blockp.X * MAP_BLOCKSIZE + MAP_BLOCKSIZE/2 * MAP_BLOCKSIZE,
+			blockp.Y * MAP_BLOCKSIZE + MAP_BLOCKSIZE/2 * MAP_BLOCKSIZE,
+			blockp.Z * MAP_BLOCKSIZE + MAP_BLOCKSIZE/2 * MAP_BLOCKSIZE);
```
There's one multiplication too much here

sfan5 commented on src/network/networkprotocol.h:
```diff
@@ -896,11 +911,13 @@ enum ToServerCommand
+		u8 count_mb
+		for count_mb:
+			v3s16 mapblock_p
+		# Following added mid of version 26
```
we're at ver 40 btw

sfan5 commented on src/network/serverpackethandler.cpp:
```diff
@@ -1841,3 +1845,58 @@ void Server::handleCommand_HaveMedia(NetworkPacket *pkt)
+	// a pre-prioritized list.
+	std::vector<WantedMapSend> wanted_map_send_queue;
+	wanted_map_send_queue.reserve(wanted_map_send_queue_len);
+	for (u32 i=0; i<wanted_map_send_queue_len; i++) {
+		WantedMapSend wms;
+		wms.type = (WantedMapSendType)pkt_in->read<u8>();
+		wms.p = pkt_in->read<v3s16>();
+		wanted_map_send_queue.push_back(wms);
+	}
```
```suggestion
	for (u32 i = 0; i < wanted_map_send_queue_len; i++) {
		wanted_map_send_queue.emplace_back();
		WantedMapSend &wms = wanted_map_send_queue.back();
		wms.type = (WantedMapSendType)pkt_in->read<u8>();
		wms.p = pkt_in->read<v3s16>();
	}
```

sfan5 commented on src/network/serverpackethandler.cpp:
```diff
@@ -1841,3 +1845,58 @@ void Server::handleCommand_HaveMedia(NetworkPacket *pkt)
+	client->setAutosendParameters(autosend_radius_map, autosend_radius_far,
+			autosend_far_weight, autosend_fov, autosend_max_total_mapblocks,
+			autosend_max_total_farblocks);
+	client->setMapSendQueue(wanted_map_send_queue);
```
this could be changed to not copy the data (`std::move`)

sfan5 commented on src/network/wms_priority.h:
```diff
@@ -0,0 +1,81 @@
+			break;
+		}
+		return priority;
+	}
```
at this line count it should probably be moved to wms_priority.c

sfan5 commented on src/server.cpp:
```diff
@@ -2404,73 +2404,113 @@ void Server::SendBlockNoLock(session_t peer_id, MapBlock *block, u8 ver,
+	size_t i;
+
+	WantedMapSendQueue(): i(0) {}
+};
```
unused

sfan5 reviewed (commented):
I looked through the code a bunch

lhofhansl commented on src/clientiface.cpp:
```diff
@@ -46,396 +45,245 @@ const char *ClientInterface::statenames[] = {
+					emerge_required = true;
 
-				if (m_occ_cull && !block_not_found &&
-						env->getMap().isBlockOccluded(block, cam_pos_nodes)) {
```
Do we rely on the client doing occlusion culling and only requested needed blocks?

lhofhansl commented:
Generally I'd love to get this (and farmaps) in!

Zughy commented:
@celeron55 friendly reminder to check other core devs reviews!

Zughy commented:
more than 30 days from the change request, no answer from the author, not here nor on IRC. Closing with adoption needed

lhofhansl commented:
Looks like @celeron55 lost interest. :)
By itself this change won't do much, but useful if we wanted to re-add farmaps.

celeron55 commented:
Well, I should note this here too: I haven't lost interest, but I have lost time. I will get back to this - but I won't promise earlier than my next summer vacation, as this summer's vacation was where I found the time for this to begin with. Unfortunately I have a very intensive job elsewhere at the moment.

lhofhansl commented:
Maybe one of us can finish this one and get it merged.

rubenwardy commented:
It's only worth doing that if far map is done as well. Otherwise you're just making the map load slower for now reason

lhofhansl commented:
It's the next summer vacation... @celeron55 feel like picking this up again?
With client block chunking MT easily works with greater viewing_ranges even without farmaps, so a better algorithm for sending blocks would be great.

I've tweaked the current one a bit in various PRs, but I think we pushed that one as far as it can go.
